Update inlinemacros.md

xeno-by opened the pull request:
Please don't comment on proposed changes until I write additional clarifying comments of my own. I'll leave a message when this pull request is ready for review.

xeno-by committed:
Update inlinemacros.md

xeno-by commented on drafts/inlinemacros.md:
```diff
@@ -25,9 +26,7 @@ We introduce a new reserved word: `inline`. `inline` can be used as a modifier f
-write such `val`s now as `inline`.  Inline values can be defined
-anywhere, but inline methods must be members of a class, trait or
-object.
+write such `val`s now as `inline`.
```
Not sure why inline methods can't be local. Is there anything that I've missed?

odersky replied:
Agreed. The previous restriction was there because it made it easier to talk about the prefix of an inline method call. Since we don't do prefixes anymore in the inline world, it can go.

xeno-by commented on drafts/inlinemacros.md:
```diff
@@ -35,8 +34,12 @@ members. Instead, every inline member becomes an overloaded alternative
-Inline definitions only exist at compile-time; no storage is allocated for them and
-no code is generated for them. That means that it is OK to have an inline member that
+If an inline definition doesn't explicitly specify its result type,
+the result type gets inferred according to the usual rules.
```
I think that allowing type inference here makes sense, especially given that I've removed the possibility of whitebox expansion at the bottom of the document. For a discussion of that, see https://github.com/odersky/sips/pull/3#discussion_r29100177.

xeno-by commented on drafts/inlinemacros.md:
```diff
@@ -35,8 +34,12 @@ members. Instead, every inline member becomes an overloaded alternative
+the result type gets inferred according to the usual rules.
+
+Inline definitions only exist at compile time; no storage is allocated for them in object layout
+and no code is generated for them in object method tables.
```
Just reworded the original formulation. One might argue that inline definitions have storage allocated for them in TASTY and _byte_code generated for them in classfiles, so I tried to make this a bit more careful. I'm not very happy with the result, though.

xeno-by commented on drafts/inlinemacros.md:
```diff
@@ -62,16 +65,19 @@ is equivalent to
+The compiler will do the following rewritings when encountering certain patterns of code
+outside the bodies of inline values and inline methods.
+When these patterns of code appear inside the aforementioned constructs, 
+they will be ignored by the inliner and processed by the compiler as usual.
```
Do we expand inline applications inside other inline methods? 

```
inline def foo() = ...
inline def bar() = foo() // (1)
bar // (2)
// when do we evaluate ...? at (1) or at (2)?
```

What about inline values?

```
inline val bar = ... // (1)
inline def foo() = bar // (2)
foo() // (3)
// when do we evaluate ...? at (1) or at (2) or at (3)?
```

```
inline def foo() = ...
inline val bar = foo() // (1)
bar // (2)
// when do we evaluate ...? at (1) or at (2)?
```

Do we expand inline methods and values inside macro expressions? 

```
inline def foo() = ...
inline def bar() = macro(foo()) // (1)
bar() // (2)
// when do we evaluate ...? at (1) or at (2)
```

These are the easier questions. There are also combinations. What if we have a standalone macro expression whose body refers to an inline val whose body calls an inline method? When do these things expand?

DarkDimius replied:
Actually, as we agreed that in order for method to become `inline` it could need to redo the overloading resolution, expansion order matters. I would say we need to stick to one decision, and as we decided that expansion happens outside-in than It feels natural to say that inline methods do not expand inside inline methods(otherwise they will lose more precise info that overload resolution could have used).

All this questions of order of expansion seem very subtle for me. I'm sure it would be a source of confusion. And could also become a source of problems, like inlining order in JVM.

xeno-by replied:
About the overloading thing, let's discuss this at https://github.com/odersky/sips/pull/3/files#r29100225.

xeno-by replied:
1) (2)
2а) (1)
2b) (1) 
3) (2)

xeno-by commented on drafts/inlinemacros.md:
```diff
@@ -62,16 +65,19 @@ is equivalent to
 method, where the prefix `prefix` and all arguments `argsI` for by-value
-parameters are side-effect free path expressions, replace the expression 
-with the method's right hand side, where parameter references are replaced 
+parameters are side-effect free path expressions or values of inline types, 
```
I think it's reasonable to allow inline values as well, because they are supposed to be side-effect free, aren't they?

DarkDimius replied:
> they are supposed to be side-effect free

If I got it right inline values are Function0's from semantic point of view, ie could have side effects.
If not, how could we check absence of side effects?

xeno-by commented on drafts/inlinemacros.md:
```diff
@@ -80,39 +86,37 @@ by-value parameters and that are not side-effect free paths. Arguments to by-nam
-of function applications under inlining: An  function call should have the same
-semantics independently on whether the function was maked `inline` or not.
+of function applications under inlining: A function call should have the same
+semantics with respect to side effects independently on whether the function was made `inline` or not.
```
A small clarification. Added `with respect to side effects`.

xeno-by commented on drafts/inlinemacros.md:
```diff
@@ -80,39 +86,37 @@ by-value parameters and that are not side-effect free paths. Arguments to by-nam
+
+   After the rewriting is complete, the result of the rewriting is forcibly upcast
+to the type of the original expression. This mechanism works exactly the same 
+as the mechanism of blackbox macro expansion in Scala 2.10+.
```
I think we should outright prohibit any type of whitebox expansion (of both inlines and macros) before we implement blackbox expansion. Whiteboxity is fundamentally incompatible with the notion of macro expansion after typer, which is so far the only reliable way of doing full-blown macro expansion (not counting the syntax-only macro annotations).

xeno-by replied:
Just to clarify for someone outside EPFL who'll be reading this. I realize that whitebox macros enable important use cases, and I'm not intending to give up on whiteboxity just yet. However, what I would personally like to do with the new macro system is to start with something that works 100% robustly and then add stuff as opposed to starting with something that is 100% cool and then thinking how to make it work.

xeno-by commented on drafts/inlinemacros.md:
```diff
@@ -80,39 +86,37 @@ by-value parameters and that are not side-effect free paths. Arguments to by-nam
+can, if necessary, be emulated by the new style of inline rewritings.
+
+2. If `prefix.f[Ts](args1)...(argsN)` refers to a partially applied inline
+method, an error is raised. Eta expansion of inline methods is prohibited.
```
I don't see how we can allow eta expansion of inline methods. Wdyt?

xeno-by replied:
See https://github.com/odersky/sips/pull/3#commitcomment-11015517

xeno-by commented on drafts/inlinemacros.md:
```diff
@@ -80,39 +86,37 @@ by-value parameters and that are not side-effect free paths. Arguments to by-nam
-     final def foo(x: Int)(y: Int = inline x + 1) = ...
-
-Then `foo(0)()` expands to `foo(0)(0 + 1)`
-
```
This is not a suggestion to remove inline default arguments, but a call for clarification. There is a number of unclear aspects of this section, mostly in how it avoid talking about side effects, and we might want to address that.

Here's how default/named expansion works in scalac:

```
def foo(x: Int)(y: Int = x.+(1)): Nothing = scala.this.Predef.???;

// compiles to
def foo(x: Int)(y: Int): Nothing = ...
<synthetic> def foo$default$2(x: Int): Int = x.+(1);
```

```
foo(0)()

// expands to:
{
  <artifact> val x$1: Int = 0;
  <artifact> val x$2: Int = Test.this.foo$default$2(x$1);
  Test.this.foo(x$1)(x$2)
}
```

We can see that the compiler takes its time to ensure that all side effects occur in order and exactly once. I'm not sure that this can be achieved with the proposed `foo(0)(0 + 1)` expansion scheme.

xeno-by commented on drafts/inlinemacros.md:
```diff
@@ -121,7 +125,7 @@ of inline types, and inline types are subtypes of their underlying types. E.g.
     42  <:  inline Int  <:  Int
 
-    (Float => inline Boolean)  <:  (inline Float) => (inline Boolean)  <:  (inline Float) => Boolean
+    (Float => inline Boolean)  <:  inline ((inline Float) => (inline Boolean)) <:  inline ((inline Float) => Boolean)
```
Looks like `inline` qualifiers were missing on the outermost `=>` types here.

xeno-by commented on drafts/inlinemacros.md:
```diff
@@ -144,6 +148,7 @@ positions:
 
         inline val x: inline T = E
 
+
```
Why do we need to silently convert from something that the user has written, i.e. `: T`, to something that they haven't, i.e. `: inline T`? Won't the type inference take care of assigning the correct type here?

DarkDimius replied:
If I got it right, we never infer if type of definition is `inline`. As `inline` has different guarantees with respect to binary compatibility and runtime behavior and subtyping.

xeno-by commented on drafts/inlinemacros.md:
```diff
@@ -165,99 +170,66 @@ positions:
  - types of abstract methods or fields
- - types of variables
- - types of class parameters
+ - types of vars
```
I think, the word `variables` is too broad here. Some people in the Scala community call `vals` variables as well.

DarkDimius replied:
mutable variables?

xeno-by commented on drafts/inlinemacros.md:
```diff
@@ -165,99 +170,66 @@ positions:
  - type parameters to methods
  - type parameters in `new` expressions or class parents.
 
-## Inline Overloading Resolution
```
I might be mistaken, but isn't this section obsoleted by the recently introduced by-value/by-name thingie?

DarkDimius replied:
Why would it?
Lets say you have methods on `List`:

```
def sum(implicit Numeric[T]): T
inline def sum(this: List[Int]): Int
```

sum is an overloaded method in this case.

xeno-by commented on drafts/inlinemacros.md:
```diff
@@ -165,99 +170,66 @@ positions:
-maybe controlled by a language import such as
-
-    import language.whiteboxMacros
-
```
See the discussion at https://github.com/odersky/sips/pull/3/files#r29100177.

xeno-by commented on drafts/inlinemacros.md:
```diff
@@ -165,99 +170,66 @@ positions:
 ## Macros
 
-A macro is an expression of the form
+A macro expression is an expression of the form
```
As agreed upon in one of the emails, changed the terminology to say `macro expression` instead of `macro`. Not only here, but everywhere in the document.

xeno-by commented on drafts/inlinemacros.md:
```diff
@@ -165,99 +170,66 @@ positions:
-
-the type of the macro itself would be
-
-    inline List[inline (inline String, T)]
```
I really like the systemicity of the transformation, as well as the duality wrt the transformation applied to the result of the macro, but I think there's something that's missing.

Let's say that along with a macro expression we have declared a helper that calculates a name of a member that we want to reference. Due to some whim, we decided to declare this helper as an inline method. Note the type of the method: `inline (inline Int, Boolean) => String`, just like in an example above (well, strictly speaking, a method has method type, but you get the idea).

```
inline def genName1(x: inline Int, b: Boolean): String = 
  if (b) ("Foo" + x) else ("Bar" + x)

inline val genName2: inline ((inline Int, Boolean) => String) = 
  (x: inline Int, b: Boolean) => <the same>

macro {
  val b: Boolean = ???
  val name1: String = genName1(42, b) // (1)
  val name2: String = genName2(42, b) // (2)
  ...
}
```

Now the example above tells us that `genName` should be seen in the macro body as `(Int, Term) => Term`, but that's not what we wanted in this case. In this case, we wanted a helper that would take a runtime value of Boolean (a runtime value at compile time) and then use it to generate a runtime value of String that we'll convert to a name.

I think that the example shows that, inside macro expressions, inline things can be useful in both lifted and unlifted forms. Therefore, I propose that we:
1) Introduce special functions: `inline def lift[T, L](x: inline T)(implicitly ev: Lift[T, L]): L = ev(x)` and `inline def unlift[L, T](x: L)(implicitly ev: Unlift[L, T]): inline T = ev(x)` that perform the conversions as necessary,
2) Expose values of type `inline T` with their types unchanged.

DarkDimius replied:
I believe that `Liftable` and `Unliftable` give a huge increase of what could be done in macros. It would be good to start with them by default and include implementations of `(Un)?Liftables` for most of classes present in standard library.

densh replied:
Not inversing annotations is inconsistent, because you are inversing evaluation semantics when you enter macro expression scope from regular runtime to compile-time.

xeno-by replied:
There can be multiple understandings of consistency. We just need to clearly define what belongs to compile-time and runtime worlds (the usefulness of this separation can be seen from the experience of Scheme and its recent development in Racket).

One definition could be that inlines belong to compile-time world, macros belong to compile-time world, statics belong to both compile-time world and runtime world, the rest is runtime (that's what I'm proposing in the comment). Once we settle on a definition (and, of course, the discussion is not closed - let's consider the alternatives), let's stay consistent with it, I agree.

xeno-by replied:
Looks like we MUST invert all the types, because of a counter-example that Martin has come up with:

```
val y: Int = ???
inline def foo(x: Int): Int = x + y
macro {
  val z = foo(42)
}
```

If we don't invert the type of `foo`, allowing to call it as `Int => Int`, then this doesn't work, because we can't evaluate 42 + y. Therefore, making type inversion optional is not only arguably unnice, but also unsound.

xeno-by commented on drafts/inlinemacros.md:
```diff
@@ -165,99 +170,66 @@ positions:
+but blocks are used most commonly). 
+
+In a macro expression, `macro` stands for a function 
+`def macro[T](body: implicit scala.meta.macros.Context => scala.meta.Term[T]): T = ???` declared in `scala.Predef`.
```
Since I wanted to discuss how we convert between lifted and unlifted representations of inline things, I've written this signature in the most trivial way - without accounting for the possibility that a macro might return something other than a term and then have it lifted by the framework. See the discussion at https://github.com/odersky/sips/pull/3/files#r29100322.

xeno-by replied:
The necessity to explicitly specify the type of `Term[T]` will be really annoying, especially since after Scala 2.11 people are accustomed to returning unadorned quasiquotes from their macros, e.g. `def impl(c: Context) = { import c.universe._; q"42" }`. 

Therefore, how about we allow inference of that `T` thing from an expected type at the typecheck site of the macro expression. E.g. `inline def foo: Int = macro { q"42" }` would infer `T` to be `Int` after the return type of the inline method.

xeno-by replied:
If a type argument to `Predef.macro` is missing, then: 1) if `pt` is specified, we use it to infer the type argument, 2) if `pt` is not specified, then it's a compilation error.

xeno-by replied:
Given that we decided that full inversion of types crossing meta boundaries is a good idea, the signature for `Predef.macro` is going to look like `def macro[MetaType, RuntimeType](body: implicit scala.meta.macros.Context => MetaType)(implicit ev: Meta[MetaType, RuntimeType]): RuntimeType = ???`.

It seems that Martin is okay with having a magic `Meta` typeclass that captures type inversion.

densh replied:
Another name idea: `meta.Inversion[From, To]`.

xeno-by commented on drafts/inlinemacros.md:
```diff
@@ -165,99 +170,66 @@ positions:
+In a macro expression, `macro` stands for a function 
+`def macro[T](body: implicit scala.meta.macros.Context => scala.meta.Term[T]): T = ???` declared in `scala.Predef`.
+Description of the functionality exposed by the `scala.meta` metaprogramming library 
+is outside of the scope of this proposal.
```
A cop out? Yes. But that's the best that I can do at the moment, and I think it's better to be honest and explicit about this.

xeno-by replied:
We could, for example, add that macro expressions are experimental until we see how we want to shape the interface between the compiler and macro bodies and until then we don't specify the interface at all (apart from some magic names that start with `scala.meta`).

xeno-by commented on drafts/inlinemacros.md:
```diff
@@ -165,99 +170,66 @@ positions:
+but local terms of type `T` are transformed into terms of type `scala.meta.Term[T]`, 
+and local types `U` are transformed into terms of type `scala.meta.Type`. In other words, 
+definitions that are statically available outside macro bodies remain available in macro bodies,
+whereas term and type parameters of enclosing inline methods become available as their representations.
```
See the discussion at https://github.com/odersky/sips/pull/3/files#r29100322.

xeno-by commented on drafts/inlinemacros.md:
```diff
@@ -165,99 +170,66 @@ positions:
+## Macro expansion
+
+During typechecking, the compiler treats macro expressions, i.e. invocations of `Predef.macro`,
+as normal method calls, typechecking macro bodies and performing type inference if necessary, but nothing else.
```
We don't need to prohibit eta expansion of `Predef.macro`, because I can't see how people could make it typecheck :)

xeno-by commented on drafts/inlinemacros.md:
```diff
@@ -165,99 +170,66 @@ positions:
+At some point in the compilation pipeline, after typechecking is complete for the entirety of the program, 
+the compiler expands macro expressions outside the bodies of inline values and inline methods.
+When macro expressions code appear inside the aforementioned constructs, 
+they will be ignored by the macro expander and processed by the compiler as usual.
```
Expansion order discussion: https://github.com/odersky/sips/pull/3/files#r29100167

xeno-by commented on drafts/inlinemacros.md:
```diff
@@ -165,99 +170,66 @@ positions:
+
+After the expansion is complete, the result of the expansion is forcibly upcast
+to the type of the original expression. This mechanism works exactly the same 
+as the mechanism of blackbox macro expansion in Scala 2.10+.
```
See the discussion at https://github.com/odersky/sips/pull/3/files#r29100177

xeno-by replied:
This formulation is almost a verbatim copy of a paragraph describing the same thing about inline expansion. Can we deduplicate somehow?

xeno-by commented:
I think I'm ready. Review @odersky @DarkDimius @vjovanov @densh.

xeno-by commented on drafts/inlinemacros.md:
```diff
@@ -80,39 +86,37 @@ by-value parameters and that are not side-effect free paths. Arguments to by-nam
 parameters are left alone.
```
I'm somewhat concerned by the fact: 1) that not only people will mistakenly write by-value parameters for inline macros and expect their trees, but 2) also that in some cases that will work. Example:

```
inline def foo(x: Int) = macro { ... x ... }
foo(42) // (1) stable, getting q"42" 
foo(M.x) // (2) stable, getting q"M.x"
foo(new C().x) // (3) not stable, getting q"x$152"
```

If a macro newbie tried to write a shiny new inline macro and only tested the first two cases, they won't even have a clue that their macro will fail on (3) and on virtually anything else.

xeno-by replied:
Therefore, how about we consistently do the introduction of temporary variables all the time? Would our optimizer handle the influx of garbage temporaries? /cc @DarkDimius

DarkDimius replied:
> Would our optimizer handle the influx of garbage temporaries?

It would. Actually optimizer itself introduces a huge amount of those during optimization and then cleans them up.
I'm supporting the idea to always introduce temporary values for call-by-value arguments.

xeno-by replied:
We agreed that this makes sense.

xeno-by commented on drafts/inlinemacros.md:
```diff
@@ -62,16 +65,19 @@ is equivalent to
 by `prefix`. 
 
-   If `prefix` is not a side-effect free path, lift it out to
+   If `prefix` is not a side-effect free path or a value of inline type, lift it out to
```
Since the idea of having `this` as an explicit parameter doesn't seem to be getting traction, how about an alternative proposal:

```
class C { self: (=> C) =>
  inline def returnMyself = macro { self } // or `macro { this }`
}
```

xeno-by replied:
Didn't look good to Martin.

xeno-by commented:
@odersky ping

xeno-by committed:
addresses some feedback

xeno-by commented:
Addressed some of the feedback, remaining points to discuss on Tuesday (/cc @vjovanov):

General:
- Relative order of expansion when we mix inline/inline, inline/macros, etc https://github.com/odersky/sips/pull/3#discussion_r29100167
- Consistently introduce temporary variables all the time? https://github.com/odersky/sips/pull/3#discussion_r29100420
- Accessors will be added as needed? https://github.com/odersky/sips/pull/3#commitcomment-11015500

Macros:
- "Macro" vs "macro expression" https://github.com/xeno-by/sips/commit/6159c7fd520c2d784c18bd3d2ab9239de8f59f0b#commitcomment-11015536
- How to inverse the types inside macro bodies? https://github.com/odersky/sips/pull/3#discussion_r29100322
- Signature for Predef.macro: https://github.com/odersky/sips/pull/3#discussion_r29100347
- Extended type inference rules for macro expressions https://github.com/odersky/sips/pull/3#discussion_r29100382
- Semantics of `this` in macros https://github.com/xeno-by/sips/commit/6159c7fd520c2d784c18bd3d2ab9239de8f59f0b#commitcomment-11015538 as well as  https://github.com/odersky/sips/pull/3#discussion_r29100425

xeno-by committed:
clarification on the order of expansion

xeno-by committed:
normalizes whitespace

xeno-by committed:
temporary variables are now created all the time, not only for unstable values

xeno-by committed:
clarification on using private/protected method in inline expansions

xeno-by committed:
consistently using macro expressions everywhere

I'm almost convinced about calling them "meta expressions",
but I need additional time to think through the consequences,
so let's use "macro expressions" for the time being.

xeno-by committed:
restores type inversion

xeno-by committed:
macros can no longer use `this` directly

xeno-by committed:
refines the signature of Predef.macro

xeno-by committed:
rewords the explanation of inversion

xeno-by committed:
describes the proposed type inference change for macro expressions

xeno-by committed:
some rewording and reformatting

xeno-by commented:
@odersky I've addressed all the feedback from our today's meetings (see https://github.com/odersky/sips/pull/3#issuecomment-98465368 for the list of questions) in the 10+ new commits provided above.

Instead of looking through changes one-by-one, I would suggest going to https://github.com/xeno-by/sips/blob/patch-1/drafts/inlinemacros.md and then reading: a) the first bullet point in `Inline Reductions` and b) the entire `Macro Expressions` section. Other changes are too minor to warrant a complete re-read.

The remaining points to discuss later (e.g. on Wednesday's inline meeting):
- The `Inline Overloading Resolution` section
- `macro { ... }` vs `meta { ... }`. I'm almost okay with this, but I need to sleep on it. Let's revisit this point one of these days to see whether the new terminology holds water.

xeno-by committed:
Term[T] => just Term

xeno-by committed:
macro expressions => meta expressions

xeno-by committed:
inlinemacros.md => inlinemeta.md

xeno-by committed:
scala.meta.Term => scala.meta.Expr

xeno-by committed:
an important clarification on the new scheme of type inference

xeno-by commented:
New day, new updates:
- Rewritten the paragraph about type inference for meta expressions.
- Renamed everything that said "macro" to say "meta". Namely: 
  - Macro expression => meta expression
  - Macro body => meta scope (I like this one)
  - Macro expansion => meta expansion (I don't like this one, but "expansion of meta expressions" is even worse - a mouthful and also too specific)
  - `scala.meta.macros.Context` => `scala.meta.semantic.Context`
  - `Predef.macro` => `scala.meta`
  - `inlinemacros.md` => `inlinemeta.md`

So again, to review my changes, I would suggest going to https://github.com/xeno-by/sips/blob/patch-1/drafts/inlinemeta.md (note that the URL has changed!) and then reading: a) the first bullet point in Inline Reductions and b) the entire Macro Expressions section.

xeno-by committed:
updates the list of authors as discussed during the meeting

xeno-by commented:
@odersky Everything's ready for review. The level thing that we discussed will be submitted as a follow-up pull request.

xeno-by committed:
consistently refers to inline methods as "methods"

Since inlinity is not first-class, we can't say that inline methods
are functions.

xeno-by committed:
don't hoist by-name arguments

xeno-by committed:
allows `this` references in macro scopes.

xeno-by committed:
regressing to level 1

An explanation of the level system (levels 1, 2, 3, and 4) with respect to
the `inline` language feature will be given in a follow-up pull request.

xeno-by committed:
includes inline values in the "Inline Reductions" section

xeno-by committed:
explicitly says that we want the meta.`package`.apply

xeno-by committed:
introduces inline parameters

xeno-by committed:
rewrites the name availability subsection once again

xeno-by committed:
inline parameters can only take literal arguments

Calls to constructors of inline classes would be nice to support as legal
values for inline parameters, but I've no idea how to then pass these
values into macro scopes.

xeno-by committed:
allows returning anything liftable from a meta scope

xeno-by commented:
@odersky As discussed before your departure, I'm closing this pull request, so that @vjovanov and I can work on writing down our findings on inline types and different levels. In case you'll be interested in seeing our progress, it'll be available at https://github.com/xeno-by/sips/blob/gh-pages/drafts/inlinemeta.md.
